File: solutions/794/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Wed Dec  8 22:58:06 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <string> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
private:
    bool checkEmpty(vector<string>& board) {
        for (auto& s : board) {
            for (auto& c : s) {
                if (c != ' ') return false;
            }
        }
        return true;
    }
    bool checkFull(vector<string>& board) {
        for (auto& s : board) {
            for (auto& c : s) {
                if (c == ' ') return false;
            }
        }
        return true;
    }
    bool checkWin(string& s) {
        if (s[0] == s[1] && s[1] == s[2]) {
            if (s[0] == 'X' || s[0] == 'O') return true;
        }
        return false;
    }
    bool shouldEnd(vector<string>& board) {
        string s = "";
        if (checkFull(board)) return true;
        else {
            for (int i = 0; i < 3; ++i) {
                if (checkWin(board[i])) return true;
                else {
                    s+=board[0][i];
                    s+=board[1][i];
                    s+=board[2][i];
                    if (checkWin(s)) return true;
                    s.clear();
                }
            }
            s+=board[0][0];
            s+=board[1][1];
            s+=board[2][2];
            if (checkWin(s)) return true;
            s.clear();
            s+=board[0][2];
            s+=board[1][1];
            s+=board[2][0];
            if (checkWin(s)) return true;
            s.clear();
        }
        return false;
    }
    bool backtracking(vector<string>& boardTemp, char cPre, vector<string>& board) {
        // base case
        if (boardTemp == board) return true;
        if (shouldEnd(boardTemp)) return false;

        // prune
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                if (boardTemp[i][j] != ' ' && boardTemp[i][j] != board[i][j]) return false;
            }
        }

        char cNext;
        if (cPre == ' ' || cPre == 'O') cNext = 'X';
        else cNext = 'O';

        // backtracking
        for (int i = 0; i < 3; ++i) {
            for (int j = 0; j < 3; ++j) {
                if (boardTemp[i][j] != ' ') continue;
                boardTemp[i][j] = cNext;
                if (backtracking(boardTemp, cNext, board)) return true;
                boardTemp[i][j] = ' ';
            }
        }

        return false;
    }
public:
    bool validTicTacToe(vector<string>& board) {
        if (checkEmpty(board)) return true;
        vector<string> boardTemp(3, "   ");
        return backtracking(boardTemp, ' ', board);
    }
};

TEST(Test794, SimpleTest) {
    vector<string> board{
        "XOX","O O","XOX"
    }; 
    Solution s; 
    EXPECT_TRUE(s.validTicTacToe(board));
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/794/main.py
from typing import List 

class Solution:
    @staticmethod
    def checkWin(s):
        if s[0] == s[1] and s[1] == s[2]:
            if s[0] == 'X' or s[0] == 'O':
                return True
        return False

    @staticmethod
    def checkFull(board):
        for s in board:
            if s[0] == ' ' or s[1] == ' ' or s[2] == ' ':
                return False
        return True

    @staticmethod
    def checkEmpty(board):
        for s in board:
            if s[0] != ' ' or s[1] != ' ' or s[2] != ' ':
                return False
        return True

    def shouldEnd(self, board):
        # check if full
        if self.checkFull(board):
            return True
        # check if any player wins
        for i in range(3):
            if self.checkWin(board[i][0]+board[i][1]+board[i][2]):
                return True
            elif self.checkWin(board[0][i]+board[1][i]+board[2][i]):
                return True
        if self.checkWin(board[0][0]+board[1][1]+board[2][2]):
            return True
        elif self.checkWin(board[0][2]+board[1][1]+board[2][0]):
            return True
        return False

    def backtracking(self, board_temp, cPre, board):
        # base case
        if board_temp == board:
            return True
        elif self.shouldEnd(board_temp):
            return False

        # prune
        for i in range(3):
            for j in range(3):
                if board_temp[i][j] != ' ':
                    if board_temp[i][j] != board[i][j]:
                        return False

        if cPre == ' ' or cPre == 'O':
            cNext = 'X'
        elif cPre == 'X':
            cNext = 'O'
        cPre = cNext

        for i in range(3):
            for j in range(3):
                if board_temp[i][j] != ' ':
                    continue
                text = list(board_temp[i])
                text[j] = cNext
                board_temp[i] = "".join(text)
                if self.backtracking(board_temp, cPre, board):
                    return True
                text = list(board_temp[i])
                text[j] = ' '
                board_temp[i] = "".join(text)

        return False

    def validTicTacToe(self, board: List[str]) -> bool:
        if self.checkEmpty(board):
            return True
        board_temp = ["   " for _ in range(3)]
        return self.backtracking(board_temp, ' ', board)

if __name__ == "__main__":
    board = ["XOX","O O","XOX"]
    s = Solution()
    assert s.validTicTacToe(board) 
